### PR TITLE
Make rcu_thread_key context-aware

### LINKED ISSUE #24581: ThreadLocalStorage leak in openssl 3.3.1
Hey when i run an application using openssl as static lib under application verifier with enabled TLS check. Appverifier finds an TLS leak in OpenSSL

I simplified my scenario to the minimum, so the code perhaps does not make much sense (as i only create and destroy openssl in a library but it shows the leak like this)

Important that the leak is detectable, the ssl code needs to run in a dll which is freed during runtime, as appverifier can only detect leaks when the binary is unloaded during runtime of the application

[OpenSSLTlsLeakExe.zip](https://github.com/user-attachments/files/15740415/OpenSSLTlsLeakExe.zip)

[dumps.zip](https://github.com/user-attachments/files/15740430/dumps.zip)
image

```

************* Preparing the environment for Debugger Extensions Gallery repositories **************
   ExtensionRepository : Implicit
   UseExperimentalFeatureForNugetShare : true
   AllowNugetExeUpdate : true
   NonInteractiveNuget : true
   AllowNugetMSCredentialProviderInstall : true
   AllowParallelInitializationOfLocalRepositories : true

   EnableRedirectToV8JsProvider : false

   -- Configuring repositories
      ----> Repository : LocalInstalled, Enabled: true
      ----> Repository : UserExtensions, Enabled: true

>>>>>>>>>>>>> Preparing the environment for Debugger Extensions Gallery repositories completed, duration 0.000 seconds

************* Waiting for Debugger Extensions Gallery to Initialize **************

>>>>>>>>>>>>> Waiting for Debugger Extensions Gallery to Initialize completed, duration 0.015 seconds
   ----> Repository : UserExtensions, Enabled: true, Packages count: 0
   ----> Repository : LocalInstalled, Enabled: true, Packages count: 41

Microsoft (R) Windows Debugger Version 10.0.27553.1004 AMD64
Copyright (c) Microsoft Corporation. All rights reserved.


Loading Dump File [C:\dumps\OpenSSLTlsLeakExe.exe_240607_153731.dmp]
User Mini Dump File with Full Memory: Only application data is available

Comment: '
*** "C:\Program Files\WindowsApps\Microsoft.SysinternalsSuite_2024.5.0.0_x64__8wekyb3d8bbwe\Tools\procdump.exe" -accepteula -ma -j "c:\dumps" 86228 524 000001FC864B0000
*** Just-In-Time debugger. PID: 86228 Event Handle: 524 JIT Context: .jdinfo 0x1fc864b0000'

Symbol search path is: cache*c:\SymbolStore;srv*c:\SymbolStore*https://msdl.microsoft.com/download/symbols
Executable search path is: 
Windows 10 Version 19045 MP (20 procs) Free x64
Product: WinNt, suite: SingleUserTS
Edition build lab: 19041.1.amd64fre.vb_release.191206-1406
Debug session time: Fri Jun  7 15:37:31.000 2024 (UTC + 2:00)
System Uptime: 3 days 7:17:06.146
Process Uptime: 0 days 0:00:11.000
..........................
Loading unloaded module list
..
This dump file has a breakpoint exception stored in it.
The stored exception information can be accessed via .ecxr.

+------------------------------------------------------------------------+
| This target supports Hardware-enforced Stack Protection. A HW based    |
| "Shadow Stack" may be available to assist in debugging and analysis.   |
| See aka.ms/userhsp for more info.                                      |
|                                                                        |
| dps @ssp                                                               |
|                                                                        |
+------------------------------------------------------------------------+

For analysis of this file, run !analyze -v
ntdll!NtWaitForMultipleObjects+0x14:
00007ff9`4b7cdc04 c3              ret
0:000> !analyze -v
*******************************************************************************
*                                                                             *
*                        Exception Analysis                                   *
*                                                                             *
*******************************************************************************

*** WARNING: Check Image - Checksum mismatch - Dump: 0x34ad9, File: 0x3da9f - c:\SymbolStore\vrfcore.dll\92B245163f000\vrfcore.dll
*** WARNING: Check Image - Checksum mismatch - Dump: 0x34ad9, File: 0x3da9f - c:\SymbolStore\vrfcore.dll\92B245163f000\vrfcore.dll
*** WARNING: Check Image - Checksum mismatch - Dump: 0x34ad9, File: 0x3da9f - c:\SymbolStore\vrfcore.dll\92B245163f000\vrfcore.dll
APPLICATION_VERIFIER_TLS_TLS_LEAK (350)
Unloading DLL that allocated TLS index that was not freed.
This stop is generated if a DLL that allocated a TLS index is being unloaded
before freeing that TLS index. To debug this stop:
$ du parameter3 - display the name of the culprit DLL;
$ .reload xxx.dll=parameter4 - reload symbols for the culprit DLL (if needed).
xxx.dll is the name of the DLL displayed in the above step;
$ u parameter2 - disassemble the code that allocated the TLS. This should
point to the function that allocated the TLS but forgot to free it
before the DLL was unloaded. 
Arguments:
Arg1: 00000000000aabba, TLS index 
Arg2: 00007fffb92f636e, Address of the code that allocated this TLS index. 
Arg3: 000001fc86f4ba4c, DLL name address. Use du to dump it. 
Arg4: 00007fffb8ee0000, DLL base address. 
*** WARNING: Unable to verify checksum for OpenSSLTlsLeakDll.dll

KEY_VALUES_STRING: 1

    Key  : AVRF.Code
    Value: 350

    Key  : AVRF.Enabled
    Value: 1

    Key  : AVRF.Exception
    Value: 1

    Key  : Analysis.CPU.mSec
    Value: 702

    Key  : Analysis.Elapsed.mSec
    Value: 3421

    Key  : Analysis.IO.Other.Mb
    Value: 0

    Key  : Analysis.IO.Read.Mb
    Value: 21

    Key  : Analysis.IO.Write.Mb
    Value: 21

    Key  : Analysis.Init.CPU.mSec
    Value: 343

    Key  : Analysis.Init.Elapsed.mSec
    Value: 2250

    Key  : Analysis.Memory.CommitPeak.Mb
    Value: 513

    Key  : Failure.Bucket
    Value: BREAKPOINT_AVRF_80000003_OpenSSLTlsLeakDll.dll!CRYPTO_THREAD_init_local

    Key  : Failure.Hash
    Value: {3bb359f6-2e62-7db2-5da9-1cfee3afc588}

    Key  : Timeline.OS.Boot.DeltaSec
    Value: 285426

    Key  : Timeline.Process.Start.DeltaSec
    Value: 11

    Key  : WER.OS.Branch
    Value: vb_release

    Key  : WER.OS.Version
    Value: 10.0.19041.1


FILE_IN_CAB:  OpenSSLTlsLeakExe.exe_240607_153731.dmp

COMMENT:  
*** "C:\Program Files\WindowsApps\Microsoft.SysinternalsSuite_2024.5.0.0_x64__8wekyb3d8bbwe\Tools\procdump.exe" -accepteula -ma -j "c:\dumps" 86228 524 000001FC864B0000
*** Just-In-Time debugger. PID: 86228 Event Handle: 524 JIT Context: .jdinfo 0x1fc864b0000

NTGLOBALFLAG:  100

APPLICATION_VERIFIER_FLAGS:  80000020

APPLICATION_VERIFIER_LOADED: 1

CONTEXT:  (.ecxr)
rax=00007ff871f96f80 rbx=00007ff80a7b5a20 rcx=0000009016b8e000
rdx=00007ff871f96f80 rsi=000001fc8626c7c8 rdi=000001fc86f4ba4c
rip=00007ff871f826ee rsp=0000009016cff590 rbp=000001fc8626c7a0
 r8=00000000000001ff  r9=00000000000001ff r10=0000000000000000
r11=00007ff871f95f00 r12=00007fffb8ee0000 r13=00007fffb92f636e
r14=00000000000aabba r15=00000000000001a1
iopl=0         nv up ei pl nz na pe nc
cs=0033  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00000202
vrfcore!VerifierStopMessageEx+0x81e:
00007ff8`71f826ee cc              int     3
Resetting default scope

EXCEPTION_RECORD:  (.exr -1)
ExceptionAddress: 00007ff871f826ee (vrfcore!VerifierStopMessageEx+0x000000000000081e)
   ExceptionCode: 80000003 (Break instruction exception)
  ExceptionFlags: 00000000
NumberParameters: 1
   Parameter[0]: 0000000000000000

PROCESS_NAME:  OpenSSLTlsLeakExe.exe

ERROR_CODE: (NTSTATUS) 0x80000003 - {EXCEPTION}  Breakpoint  A breakpoint has been reached.

EXCEPTION_CODE_STR:  80000003

EXCEPTION_PARAMETER1:  0000000000000000

STACK_TEXT:  
00000090`16cff590 00007ff8`0a787a53     : 000001fc`86f4ba4c 00007ff8`0a7838b9 000001fc`86f5adb0 00007fff`b8ee0000 : vrfcore!VerifierStopMessageEx+0x81e
00000090`16cff8f0 00007ff8`0a78738f     : 000001fc`8626c7a0 000001fc`86f5adb0 00000000`00952000 00000000`00952000 : vfbasics!AVrfpCheckAndFreeDllLeakedTlsSlots+0x7b
00000090`16cff940 00007ff9`4b809800     : 000001fc`86f5adb0 00000000`00000000 000001fc`86f5ae50 000001fc`86363130 : vfbasics!AVrfpDllUnloadCallback+0x8f
00000090`16cff9a0 00007ff9`4b7efc9a     : 000001fc`86f5adb0 000001fc`86f5ae50 000001fc`86363130 00000000`00000000 : ntdll!AVrfDllUnloadNotification+0x9c
00000090`16cff9e0 00007ff9`4b73fd0a     : 000001fc`86363130 000001fc`86363130 00000090`16cffae0 00000000`00000000 : ntdll!LdrpUnloadNode+0x557d2
00000090`16cffa30 00007ff9`4b73fc84     : 00000000`00000000 00000090`00000001 00000000`00000000 00000000`00000000 : ntdll!LdrpDecrementModuleLoadCountEx+0x72
00000090`16cffa60 00007ff9`4907987e     : 00007fff`b8ee0000 000001fc`00000009 000001fc`86f5adb0 00000000`00000000 : ntdll!LdrUnloadDll+0x94
00000090`16cffa90 00007ff7`b6e4184f     : 00000000`00000000 00000000`00000000 00007580`b07559c3 00007fff`b8ee0000 : KERNELBASE!FreeLibrary+0x1e
00000090`16cffac0 00007ff7`b6e41d59     : 00002e05`00000001 00007ff8`051422a3 00000000`00000000 00007ff7`b6e4251d : OpenSSLTlsLeakExe!main+0x3f
00000090`16cffc00 00007ff7`b6e41c02     : 00007ff7`b6e49000 00007ff7`b6e49220 00000000`00000000 00000000`00000000 : OpenSSLTlsLeakExe!invoke_main+0x39
00000090`16cffc50 00007ff7`b6e41abe     : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : OpenSSLTlsLeakExe!__scrt_common_main_seh+0x132
00000090`16cffcc0 00007ff7`b6e41dee     : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : OpenSSLTlsLeakExe!__scrt_common_main+0xe
00000090`16cffcf0 00007ff9`4b417344     : 00000090`16b8d000 00000000`00000000 00000000`00000000 00000000`00000000 : OpenSSLTlsLeakExe!mainCRTStartup+0xe
00000090`16cffd20 00007ff9`4b7826b1     : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : kernel32!BaseThreadInitThunk+0x14
00000090`16cffd50 00000000`00000000     : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : ntdll!RtlUserThreadStart+0x21


STACK_COMMAND:  ~0s; .ecxr ; kb

FAULTING_SOURCE_LINE:  %PATH%\crypto\threads_win.c

FAULTING_SOURCE_FILE:  %PATH%\crypto\threads_win.c

FAULTING_SOURCE_LINE_NUMBER:  511

FAULTING_SOURCE_CODE:  
   507:     return (*lock == ONCE_DONE);
   508: }
   509: 
   510: int CRYPTO_THREAD_init_local(CRYPTO_THREAD_LOCAL *key, void (*cleanup)(void *))
>  511: {
   512:     *key = TlsAlloc();
   513:     if (*key == TLS_OUT_OF_INDEXES)
   514:         return 0;
   515: 
   516:     return 1;


SYMBOL_NAME:  OpenSSLTlsLeakDll!CRYPTO_THREAD_init_local+1e

MODULE_NAME: OpenSSLTlsLeakDll

IMAGE_NAME:  OpenSSLTlsLeakDll.dll

FAILURE_BUCKET_ID:  BREAKPOINT_AVRF_80000003_OpenSSLTlsLeakDll.dll!CRYPTO_THREAD_init_local

OS_VERSION:  10.0.19041.1

BUILDLAB_STR:  vb_release

OSPLATFORM_TYPE:  x64

OSNAME:  Windows 10

FAILURE_ID_HASH:  {3bb359f6-2e62-7db2-5da9-1cfee3afc588}

Followup:     MachineOwner
---------

0:000> du 000001fc86f4ba4c
000001fc`86f4ba4c  "OpenSSLTlsLeakDll.dll"
0:000> .reload OpenSSLTlsLeakDll.dll=00007fffb8ee0000
0:000> u 00007fffb92f636e
*** WARNING: Unable to verify checksum for OpenSSLTlsLeakDll.dll
OpenSSLTlsLeakDll+0x41636e [%PATH%\crypto\threads_win.c @ 511]:
00007fff`b92f636e 488b4c2430      mov     rcx,qword ptr [rsp+30h]
00007fff`b92f6373 8901            mov     dword ptr [rcx],eax
00007fff`b92f6375 488b442430      mov     rax,qword ptr [rsp+30h]
00007fff`b92f637a 8338ff          cmp     dword ptr [rax],0FFFFFFFFh
00007fff`b92f637d 7504            jne     OpenSSLTlsLeakDll!CRYPTO_THREAD_init_local+0x33 (00007fff`b92f6383)
00007fff`b92f637f 33c0            xor     eax,eax
00007fff`b92f6381 eb05            jmp     OpenSSLTlsLeakDll!CRYPTO_THREAD_init_local+0x38 (00007fff`b92f6388)
00007fff`b92f6383 b801000000      mov     eax,1


```

### PR DESCRIPTION
**Pull request by nhorman**
Currently, rcu has a global bit of data, the CRYPTO_THREAD_LOCAL object to store per thread data.  This works in some cases, but fails in FIPS, becuase it contains its own copy of the global key.

So
1) Make the rcu_thr_key a per-context variable, and force
   ossl_rcu_lock_new to be context aware

2) Store a pointer to the context in the lock object

3) Use the context to get the global thread key on read/write lock

4) Use ossl_thread_start_init to properly register a cleanup on thread
   exit

5) Fix up missed calls to OSSL_thread_stop() in our tests

(cherry picked from commit 24d16d3a1915a06a2130385a87de9a37fc09c4b9)

Fixes #24581

### CODE REVIEW

File: crypto/threads_pthread.c [hunk 1]
```diff
@@ -403,22 +386,31 @@ static struct rcu_qp *get_hold_current_qp(struct rcu_lock_st *lock)
     return &lock->qp_group[qp_idx];
 }
 
+static void ossl_rcu_free_local_data(void *arg)
+{
+    OSSL_LIB_CTX *ctx = arg;
+    CRYPTO_THREAD_LOCAL *lkey = ossl_lib_ctx_get_rcukey(ctx);
+    struct rcu_thr_data *data = CRYPTO_THREAD_get_local(lkey);
+    OPENSSL_free(data);
```
Comment (mattcaswell) [inline crypto/threads_pthread.c:394]: Previously in `free_rcu_thr_data` we also did `CRYPTO_THREAD_set_local(&rcu_thr_key, NULL)`. Should we do the equivalent here?

File: crypto/threads_win.c [hunk 2]
```diff
@@ -205,22 +184,32 @@ static ossl_inline struct rcu_qp *get_hold_current_qp(CRYPTO_RCU_LOCK *lock)
     return &lock->qp_group[qp_idx];
 }
 
+static void ossl_rcu_free_local_data(void *arg)
+{
+    OSSL_LIB_CTX *ctx = arg;
+    CRYPTO_THREAD_LOCAL *lkey = ossl_lib_ctx_get_rcukey(ctx);
+    struct rcu_thr_data *data = CRYPTO_THREAD_get_local(lkey);
+    OPENSSL_free(data);
```
Comment (mattcaswell) [inline crypto/threads_win.c:192]: Same comment as per the pthreads case (should we set the data to NULL)

File: test/threadstest.c [hunk 3]
```diff
@@ -435,9 +435,7 @@ static int _torture_rcu(void)
     writer2_done = 0;
     rcu_torture_result = 1;
 
-    rcu_lock = ossl_rcu_lock_new(1);
-    if (!rcu_lock)
-        goto out;
+    rcu_lock = ossl_rcu_lock_new(1, NULL);
```
Comment (mattcaswell) [inline test/threadstest.c:438]: Why do we no longer test if `rcu_lock` is NULL?

File: test/threadstest.h [hunk 4]
```diff
@@ -65,6 +65,7 @@ static void *thread_run(void *arg)
     *(void **) (&f) = arg;
 
     f();
+    OPENSSL_thread_stop();
```
Comment (mattcaswell) [inline test/threadstest.h:68]: Why is this needed? Don't we detect the thread stopping automatically anyway?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: test/threadstest.c (hunk 3 above)
Comment (nhorman) [reply test/threadstest.c:438]: because I'm an idiot.  will fix

File: test/threadstest.h (hunk 4 above)
Comment (nhorman) [reply test/threadstest.h:68]: I'm digging back in my mental history here, but IIRC there was a problem with the thread_run mechanism that required it.  I'll have to go back and look at the details though

Comment (nhorman) [reply test/threadstest.h]: I took this example from the DRBG test, but yeah, it appears not needed

Comment (nhorman) [reply test/threadstest.h]: Actually, I take that back, if I remove the thread_stop here I get memory leaks in the threads_test because the thread local data is never freed.

File: crypto/threads_none.c [hunk 5]
```diff
@@ -23,7 +23,7 @@ struct rcu_lock_st {
     struct rcu_cb_item *cb_items;
 };
 
-CRYPTO_RCU_LOCK *ossl_rcu_lock_new(int num_writers)
+CRYPTO_RCU_LOCK *ossl_rcu_lock_new(int num_writers, OSSL_LIB_CTX *ctx)
```
Comment (paulidale) [inline crypto/threads_none.c:26]: `ossl_unused` before the `ctx` parameter.

File: crypto/threads_pthread.c [hunk 6]
```diff
@@ -403,22 +386,32 @@ static struct rcu_qp *get_hold_current_qp(struct rcu_lock_st *lock)
...
```
Comment (paulidale) [inline crypto/threads_pthread.c:394]: NIT: blank line above

**Review by paulidale — APPROVED** [review]
A couple of nits, but otherwise looks okay.

**Review by mattcaswell — APPROVED** [review]
Since this is a backport we should probably apply some of the fixups applied here to the master branch too.

**Comment by nhorman** [comment]
I don't think any of them are particularly critical, but as a general principle, yes.  I imagine I would do that in a separate PR though, yes?

**Comment by mattcaswell** [comment]
> I imagine I would do that in a separate PR though, yes?

Yes

**Comment by nhorman** [comment]
Tracking that work in https://github.com/openssl/openssl/issues/24670 @mattcaswell

**Comment by t8m** [comment]
IMO it is absolutely necessary to have the fixups on all relevant branches (including master branch) as we do not want to regress.

**Comment by t8m** [comment]
If this cherry-picks cleanly, it does not need a separate PR.

**Comment by nhorman** [comment]
do we allow forward cherry-picks like that?

**Comment by mattcaswell** [comment]
> do we allow forward cherry-picks like that?

Normally we allow cherry-picks like that for all branches that the PR has labelled. All approvals occurred *after* the master branch label was applied to this PR - so normally that would be ok. However, the relevant commits are *fixup* commits which you cannot push directly to master (our pre-receive hooks will reject them) - so at least the commit description would have to be modified. That's a grey area.

**Comment by t8m** [comment]
@nhorman could just rebase this to reword the commit messages for the 2 fixup commits (and possibly squash them into a one) to make them normal commits. Then when cherry picking on merge we would just cherry pick the relevant commit(s).

**Comment by nhorman** [comment]
If everyone is comfortable with that, yeah, I can do a rebase

**Comment by nhorman** [comment]
rebased to squash the fixups and rename them to pass the server side hooks.  no code changes.  @mattcaswell @t8m please reconfirm approval

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the 3.3 branch and master branch (only the second commit). Thank you.
